architecture/2024-08-resource-types-registration.md

```diff
@@ -0,0 +1,580 @@
+
+---
+
+These new resource types are scoped to the Radius plane. 
+
+> eg: `/planes/radius/local/providers/System.Resources/resourceProviders`
+
+The rationale for this is to support registration/unregistration per-tenant of Radius rather than per-installation.
+
+#### Resource Type Queries
+
+The proposal also defines two new **non-resource** APIs:
+
+- `GET /planes/radius/{plane}/providers/`: list all resource providers.
+- `GET /planes/radius/{plane}/providers/{provider}` shows a specific resource provider.
+
+These APIs are optimized for queries and are consistent with similar APIs in ARM. To understand why this is important, remember that the OpenAPI schema for a resource type can be quite large. Some functionality in Radius will need to query resource types frequently, consider the need for the `rad recipe register` to validate a resource type. Building a separate API for queries allows us to optimize the behavior.
+
+These APIs can also be implemented consistently across all planes in the future. For example Azure doesn't support user-defined types, but it does support looking up the set of resource types.
+
+Lastly, there will still be some system-defined types. For example `System.Resources/resourceProviders` is a system defined type. The query API can return these, but the `System.Resources/resourceProviders` cannot. Imagine for a minute what would happen if users could remove the capability to specify a type.
+
+**Open Question:** Is *query* the right term for this? The term will show up in docs and in the TypeSpec + generated client + our code. Let's pick the term that's least confusing to us. I'm open to suggestions but I'll just roll with *query* unless there's a better idea.
+
+#### Locations
+
+UCP is a multi-location control-plane like the control-plane that inspired it (ARM). We don't use UCP/Radius in any multi-location scenarios right now but it's important we don't design APIs that lock us into a single-region design. ARM has a mature multi-location design but doesn't expose to users the ability to register types. However we can learn a lot from the scenarios supported by ARM for rolling out new API versions and new types.
+
+- Resource providers (non-declarative) are deployed per-location.
+- New functionality (API versions + types) is rolled out per-location.
+
+Some examples:
+
+> Resource provider `Contoso.Example` is only available in the `westus-1` location. UCP must reject creation requests to other locations.
+
+> Resource type `Contoso.Example/widgets` is available in all locations, but new API version `2024-08-01-preview` is only in the `eastus-2` location. UCP should block all requests for unsupported API version x location pairs.
+
+The design proposal includes axes for configuring location affinity:
+
+- `System.Resources/resourceProviders` includes a list of supported locations with an optional `address` in each location for the resource provider.
```

Comment:
Are the available locations going to be an array for a resource type?

Reply:
Please see the API design.